servant-server/test/Servant/Server/Internal/RoutingApplicationSpec.hs

```diff
@@ -40,6 +51,36 @@ simpleRun d = fmap (either ignoreE id) . try $
   where ignoreE :: SomeException -> ()
         ignoreE = const ()
 
+-------------------------------------------------------------------------------
+-- Combinator example
+-------------------------------------------------------------------------------
+
+-- | This data types writes 'sym' to 'delayedTestRef'.
+data Res (sym :: Symbol)
+
+instance (KnownSymbol sym, HasServer api ctx) => HasServer (Res sym :> api) ctx where
+    type ServerT (Res sym :> api) m = IORef (Maybe String) -> ServerT api m
+    route Proxy ctx server = route (Proxy :: Proxy api) ctx $
+        server `addBodyCheck` check
+      where
+        sym  = symbolVal (Proxy :: Proxy sym)
+        check = do
+            liftIO $ writeIORef delayedTestRef (Just sym)
+            _ <- register (writeIORef delayedTestRef Nothing)
+            return delayedTestRef
+
+type ResApi = "foobar" :> Res "foobar" :> Get '[PlainText] T.Text
+
+resApi :: Proxy ResApi
+resApi = Proxy
+
+resServer :: Server ResApi
+resServer ref = liftIO $ fmap (maybe "<empty>" T.pack) $ readIORef ref
+
+-------------------------------------------------------------------------------
+-- Spec
+-------------------------------------------------------------------------------
```

Comment:
I definitely don't insist on it, but it perhaps would be nice to have a test where the combinator allocating the resource forks a thread, and that thread waits on an MVar (with `readMVar`). Then the test checks (a) that before it `putMVar`s, the resource is not yet deallocated, and (b) after it `putMVar`s, the resource is deallocated.

What I have in mind is forking jobs that are longer running, which I imagine would be commonly used in conjunction with resources, especially if we make `Handler` a `MonadResource`.